### PR TITLE
biquadratic solving; Poly.free_symbols; ltrim

### LINKED ISSUE #12815: left trim error while solving system of equations
Regular solve runs into a problem with the following
```
>>> eqns[:2]
[x0**2 - x0, x0*x1 - x1]
>>> solve(_)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\solvers\solvers.py", line 1071, in solve
    solution = _solve_system(f, symbols, **flags)
  File "sympy\solvers\solvers.py", line 1748, in _solve_system
    result = solve_poly_system(polys, *symbols)
  File "sympy\solvers\polysys.py", line 45, in solve_poly_system
    return solve_biquadratic(f, g, opt)
  File "sympy\solvers\polysys.py", line 86, in solve_biquadratic
    q = q.ltrim(-1)
  File "sympy\polys\polytools.py", line 634, in ltrim
    if monom not in terms:
sympy.polys.polyerrors.PolynomialError: can't left trim Poly(x0*x1 - x1, x0, x1,
 domain='ZZ')
>>> eqs
[x0**2 - x0, x0*x1 - x1]
```
Manually solving gives
```
>>> solve(eqs, manual=True)
[{x0: 1}, {x0: 0, x1: 0}]
```

### LINKED ISSUE #12846: Poly(x,x,y).free_symbols -> {x, y} instead of just {x}
No free symbols of generators that don't appear in the expression of the polynomial should appear in the set of free symbols.

```
def free_symbols(poly):
 free = set()
 for i in range(len(poly.gens)):
    for m in poly.monoms():
        if i in m:
            free |= poly.gens[i].free_symbols
            break
 return free | poly.free_symbols_in_domain  # not sure about the domain part....
```

### LINKED ISSUE #12851: eliminate univariate equations (recursively) before solving
Though this is blocked on #12815, the following should both give the first answer:

```
>>> solve((x*y - y,x**2 - x))
[{x: 1}, {x: 0, y: 0}]
>>> solve((x*y - x,y**2 - y))
[{x: 0, y: 0}, {x: 0, y: 1}]
```

### PR DESCRIPTION
**Pull request by smichr**
- closes #12851
- closes #12815

Both of the following now recognize infinite solutions and neither raises an ltrim error
```
s1 = (x*y - y, x**2 - x)  
assert solve(s1) == [{x: 1}, {x: 0, y: 0}]  
s2 = (x*y - x, y**2 - y)  
assert solve(s2) == [{y: 1}, {x: 0, y: 0}]  
```
- closes #12846

dummy symbols are not included in Poly.free_symbols anymore
```
Poly(x + y, x, z).free_symbols now gives {x, y} instead of {x, y, z}
```

### CODE REVIEW

**Comment by smichr** [comment]
ping @jksuom -- any chance you could give this a once over? It's a quick review, I believe.

File: sympy/solvers/polysys.py [hunk 1]
```diff
@@ -83,9 +80,16 @@ def solve_biquadratic(f, g, opt):
     x, y = opt.gens
 
     p = Poly(p, x, expand=False)
-    q = q.ltrim(-1)
-
     p_roots = [ rcollect(expr, y) for expr in roots(p).keys() ]
+    if not all(y in i.free_symbols for i in p_roots):
+        # then the solution won't depend on y
```
Comment (jksuom) [inline sympy/solvers/polysys.py]: The solution need not depend on y. Consider `(x**2 + y**2 - 2, y**2 - 1)`.

Comment (smichr) [inline sympy/solvers/polysys.py]: I see. Please see more precise detection and added test, @jksuom .

File: sympy/solvers/polysys.py [hunk 2]
```diff
@@ -83,9 +80,15 @@ def solve_biquadratic(f, g, opt):
     x, y = opt.gens
 
     p = Poly(p, x, expand=False)
-    q = q.ltrim(-1)
-
     p_roots = [ rcollect(expr, y) for expr in roots(p).keys() ]
+    if len(p_roots) == 1 and y not in p_roots[0].free_symbols:
+        # this is not a zero-dimensional system
```
Comment (jksuom) [inline sympy/solvers/polysys.py]: It seems that this test is not reliable. Consider `(x**2 + y**2 - 1, y**2 - 1)`. I'd detect systems of non-zero dimension by checking if `gcd(f, g)` is constant or not.

**Comment by smichr** [comment]
@jksuom, it would just kill me to use gcd when the answer is staring at us in p_roots. :-)

File: sympy/solvers/polysys.py [hunk 3]
```diff
@@ -83,9 +80,17 @@ def solve_biquadratic(f, g, opt):
     x, y = opt.gens
 
     p = Poly(p, x, expand=False)
-    q = q.ltrim(-1)
+    p_roots = roots(p)
+    k, m = list(zip(*p_roots.items()))
+    if sum(m) == 1 and y not in k[0].free_symbols:
+        # this is not a zero-dimensional system
```
Comment (jksuom) [inline sympy/solvers/polysys.py]: I cannot think of a way how this could be derived from p_roots. Consider `(x**2 + y**2 - 1, x**2 - x + y**2 - 1)`.

**Comment by smichr** [comment]
Would it just be better to handle the non-zero-dimensional case rather
than trying to avoid it? Instead of returning with SolveFailed at that point the
solution `[{y: q1}, {x: p1, y: q2}]` (where q1 is the value that makes p.subs(q1) == 0)
could be returned.

**Comment by jksuom** [comment]
It is probably possible to find the positive dimensional solutions as well. Gröbner bases are not unique but `groebner` does a lot of what is needed for the gcd. This is how I'd try to approach the solution after finding `p, q = G = groebner([f, g])` (assuming that `len(G) == 2`).

If `q` does not involve `x`, then up to 4 values of `y` are solved from `q`. Then, for each `y`, one or two values of `x` are solved from `p` except when `p` becomes a constant after substitution, in which case either no value or any value of `x` is a solution.

If `q` does involve `x`, then the number of solution is infinite. In this case it should be possible to solve `x` from `q` as a linear (possibly constant) function of `y`.

If `len(G) == 1`, then the infinite number of solutions are found by solving the sole polynomial of `G` (which is `gcd(f, g)`).

**Comment by smichr** [comment]
The problem with handling non zero dimensional systems is that this would break the API for the return value. It is assumed right now that `solve_biquadratic` will return (x, y) solutions for the system of equations. There are already some exits when these conditions don't exist. I have added two more exits for when the system is not zero dimensional. Can this be committed as is or do you see a way around this?

**Comment by jksuom** [comment]
If only the 0-dimensional case is to be handled, then I think this could do the job. I only wonder if there should be some justification for the dimensionality conditions to help maintain the code.

**Comment by smichr** [comment]
> wonder if there should be some justification for the dimensionality conditions to help maintain the code

I'm not sure what you have in mind. Could you elaborate?

**Comment by jksuom** [comment]
The function is expected to return the solutions of the system (f, g) if they are finite in number (i.e., the system is 0-dimensional) and raise  SolveFailed otherwise.

The decision between the two cases is easy if d = gcd(f, g) is known: the system is 0-dimensional if and only if d is constant (i.e., f and g are relatively prime).

Recognizing the 0-dimensional case is more tedious if only G = groebner([f, g]) is known. I cannot think of any other way than looking at all cases one by one. Let p = G[0], and also q = G[1] if len(G) == 2.

If len(G) == 1, then p == d.
- If deg(p) == 0, the system is 0-dimensional (but there are no solutions).
- If deg(p) > 0, the system is not 0-dimensional.

If len(G) == 2, it is harder to find d. It is only known that d divides both p and q since they are in the ideal generated by f and g. It is also known that it must be either constant or linear, for if it were quadratic then f and g were its constant multiples, and we would have len(G) == 1. Moreover, if d is linear, then p and q are products dp' and dq' where p' and q' are linear and q' has no x. Finally, q is a polynomial in y alone unless its factor d has x.
- If q has x, then the system is not 0-dimensional (deg(d) > 0).
- If q does not have x, then d is either constant or a polynomial in y alone,
    - if d is constant, the system is 0-dimensional.
    - If d is is not constant, then the system is not 0-dimensional. In this case, p is the product of d and another linear polynomial p'. If p' does not have x, then both p and q are polynomials in y alone, in which case (p, q) cannot be a Gröbner basis. (A Gröbner basis of polynomials in a single variable has only one element.) If p' has y then p has y^2, which is also impossible in a Gröbner basis where q is a linear or quadratic polynomial of y. Hence p' is a linear polynomial of x alone.

The last case is the most problematic one. It is the one where p has a single solution which is independent of y (namely the root of p') but still p has y in the factor d. However, this does not prove that there could not be 0-dimensional systems having the same properties. (More work would be needed, and that might not be worth the trouble.)

**Comment by smichr** [comment]
OK, so let's just (as you said) test the gcd and be done with it.

File: sympy/solvers/polysys.py [hunk 4]
```diff
@@ -79,13 +77,18 @@ def solve_biquadratic(f, g, opt):
     if len(G) != 2:
         raise SolveFailed
 
+    d = f.gcd(g)
+    if not d.as_expr().is_constant(x, y):
```
Comment (jksuom) [inline sympy/solvers/polysys.py]: Perhaps `not d.is_ground` could be used.

**Comment by jksuom** [comment]
This looks good. Merging...

**Pull request merged**